### DISCUSSION TITLE
I have no idea how to build an apk. Is it is using gradlew.bat?

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#1258] [Q&A]
hi,
I spent a couple of ours googling, and searching in forums how to build an apk to test the game on my android. I am just using axmol defaults projects to keep it simple. I did manage to see the error outputs I was getting, first it was that I need java 11. I installed last java (version 20 i think) and that solved that error. So I assume it is java 11 or higher.
Then I got a Ninja not found in Path, so after some reasearch I fixed that error by downloading ninja.exe and put it in the PATH environment. So finally I run gradlew.bat in the folder proj.android and I said that It build successfully in 1 second..
soooooo....what just happened? Is gradlew.bat suppose to build the apk without the need of setting android studio? If it built the .apk where is it?! XD. I searched all the axmol project but didn't find any apk.
I am sure that is pretty basic but can someone point me if what I am doing has nothing to do with building an .apk?

thanks

**Comment by halx99** [comment]
do you read: https://github.com/axmolengine/axmol/blob/dev/docs/DevSetup.md

Just goto root of axmol, then run follow powershell script, will build HelloCpp apk for you, it will auto setup android sdk, ndk, jdk, ninja for you

`pwsh ./build.ps1 -p android -a arm64`

**Comment by rudiHammad** [comment]
hi,
yes I read it but I was on a previous axmol version that didn't have the build.ps1 file. 
I have updated and I am following the guide now. The problem that I am getting is that when I run that command I get the error: Install cmdlinetools version: 7.0 fail
I am googling to see what that is

**Comment by halx99** [comment]
does the cmdlines package download succeed?

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rudiHammad** [reply]
I opened android studio and downloaded it from there. So i guess? 
But I am getting also getting a previous error:  Cannot find path 'D:\ndk' because it does not exist
Here is a capture
![error](https://github.com/axmolengine/axmol/assets/25636976/55a0057b-8417-4f6a-9866-ecced7aed302)

**Comment by rudiHammad** [comment]
plus, I just checked my environment vasriables and I have the variable NDK_ROOT as C:\Program Files\cocos2d_dependencies\android-ndk-r23b-windows\android-ndk-r23b
So I guess  that having cocos2dx installed is messing some of my variables

**Comment by halx99** [comment]
the ndk folder doesn't exist in sdk_root, will fix it later

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rudiHammad** [reply]
okey. So it has nothing to do with the error?
because I believe it is setup correctly
![aaa](https://github.com/axmolengine/axmol/assets/25636976/bec2cf99-f28a-4aad-8e02-a3662535a9a1)

**Reply by halx99** [reply]
sorry, not correct, the script has a bug: always search in `ANDROID_HOME`

**Comment by halx99** [comment]
you can create a empty folder ndk in your sdk_root

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by rudiHammad** [comment]
hmmm...I tried that. No luck.
I am gonna keep trying. Thanks for your help

**Comment by rudiHammad** [comment]
I tried to build on windows in case it was easier with pwsh .\build.ps1 -p winuwp but more errors; 

`build1k: BUILD_ALL_OPTIONS=--target HelloCpp --config Debug --parallel, Count=5
MSBuild version 17.5.0+6f08c67f3 for .NET Framework
MSBUILD : error MSB1009: Project file does not exist.
Switch: HelloCpp.vcxproj`

I think I am about to give up and swtich to unity. I can't investe time in an engine if I can't deploy to windows, or android. It is probably me, but I think I have all in order and I still keep getting errors. Anyway, it is quite discouraging because Axmol is pretty awesome but I think I need to be sure that I won't have problems deploying my games in the future.

**Comment by halx99** [comment]
for uwp, do you install uwp workflow for visual studio 2022?

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rudiHammad** [reply]
Yes, I think I did. But it's okey. I am pause it for now, no worries. I don't time right now to keep trying. But I might check it again later.

**Comment by aismann** [comment]
D:\__GIT\axmol>pwsh ./build.ps1 -p android -a arm64
build1k: PowerShell 7.3.5
build1k:
Name                           Value
----                           -----
d                              D:\__GIT\axmol\templates\cpp-template-default\proj.android
p                              android
setupOnly                      False
prefix                         D:\__GIT\axmol\tools\external
dll                            False
xc                             {}
cc
xt                             gradle
xb                             {--config, Debug}
winsdk
a                              arm64


build1k: proj_dir=D:\__GIT\axmol, prefix=D:\__GIT\axmol\tools\external
build1k: Found suitable installed glslcc: D:\__GIT\axmol\tools\external\glslcc\glslcc.exe, version: 1.7.6
build1k: The installed cmake: D:\Program Files\CMake\bin\cmake.exe, version: 3.24.1 not match required: 3.26.4+
build1k: Using cmake: D:\__GIT\axmol\tools\external\cmake-3.26.4-windows-x86_64\bin\cmake.exe, version: 3.26.4
build1k: Looking require r23c+ in env:ANDROID_HOME=D:\dev\adt
build1k: Looking require r23c+ in env:ANDROID_SDK_ROOT=C:\Users\dreif\AppData\Local\Android\Sdk
build1k: Not found cmdlinetools, needs install: 7.0
build1k: Installing cmdlinetools version: 7.0 ...
build1k: Downloading https://dl.google.com/android/repository/commandlinetools-win-9477386_latest.zip to D:\__GIT\axmol\tools\external\commandlinetools-win-9477386_latest.zip ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (35) schannel: next InitializeSecurityContext failed: Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.
Exception: D:\__GIT\axmol\1k\build1k.ps1:585
Line |
 585 |  …             throw "Install cmdlinetools version: $sdkmanager_ver fail …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Install cmdlinetools version: 7.0 fail

**Reply by rudiHammad** [reply]
okey, I got it working for android doing whatyou did
first
`.\setup.ps1`
and then
`.\build.ps1 -p android -a arm64`

A couple of notes that may help others (correctly if I am wrong please): it has to be  in the powershell with the hability to run scripts.
I was getting a restriction error that I solved by allowing the powersheel to run scripts. A quick search on google shows you how.
And second, don't forget the .\  in the commands

Cool, it builds for android now. Thx.

Now I'll keep trying to build on windows.

**Comment by aismann** [comment]
After some manual "downloads" it works here also. (pwsh .\build.ps1 -p android -a arm64)
@halx99 
Thanks.

**Reply by halx99** [reply]
latest commits should fix the android sdk setup problem on plain new machine (without any tools for android, jdk, ninja, cmdlinetools, sdk,ndk)

**Reply by aismann** [reply]
@halx99 

Maybe a check on `setup.ps1` about the PowerShell version and a link (better download/install Yes/No) will be great too
```
C:\Users\....\axmol>setup.ps1
build1k: PowerShell 5.1.19041.3031
...
build1k: setup successfully.
```

**Comment by rudiHammad** [comment]
okey, still trying to build for windows. I followed guide of [Windows workflow] Compile engine once, link prebuilt for project.
It did created HelloCpp.exe. That is the same as building from visual studio right? 
Then I run HelloCpp.exe, and the guide then says run.bat. So I do and I get this

`Entering run.bat directory: D:\CHIMERA STUDIO\CHIMERA GAMES\learningSeries\HelloCpp\`
`The system cannot find the path specified.`
`Please input Build config(Debug,Release,MinSizeRel,RelWithDebInfo):`

so I type _debug_ and I get the error of the image
![Capture](https://github.com/axmolengine/axmol/assets/25636976/5cdddddb-0004-4597-8160-4101e7017f78)

Is it possible that the white space between CHIMERA STUDIO is causing the error.

**Reply by aismann** [reply]
> Is it possible that the white space between CHIMERA STUDIO is causing the error.

Yes!
Best way is: rename it (like: D:\CHIMERA_STUDIO\CHIMERA_GAMES

Anyhow:
The build is working well (`It did created HelloCpp.exe`)

**Reply by rudiHammad** [reply]
aahhhgg... but I can't do that now because I have many other dependencies not related to game dev that would be a pain to refactor.
The result of that isn't the same than going to visual studio and just select the configuration debug and release and build? if it is the same I prefer to keep doing from visual studio.

This has nothing to do with deploying the game on windows right? which is done through
`pwsh .\build.ps1 -p winuwp`  . Is that correct or am I misunderstanding things?
At this point I can run HelloWorld.exe, in my computer, but of course, if I want someone to test my game, sending him that file won't work because of its dependencies right? 
To be able to send a game to someone I have to run `pwsh .\build.ps1 -p winuwp` ? which is what is not working for me. 
Can you run that without problems?

**Comment by aismann** [comment]
> I have to run `pwsh .\build.ps1 -p winuwp` ? which is what is not working for me. Can you run that without problems?

why not start with a new folder (without spaces in the path) ?
and than use only:
`pwsh .\build.ps1`

**Reply by rudiHammad** [reply]
I can confirm that the space is causing the issue. I did create a new project without spaces and this now works
` Please input Build config(Debug,Release,MinSizeRel,RelWithDebInfo): Debug` and the run.bat too. So that's a progress since this morning!
But `pwsh .\build.ps1 -p winuwp` is still not working. Here is the error:

```Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\testingGame\HelloCpp> .\build.ps1 -p winuwp
build1k: PowerShell 5.1.19041.3031
build1k:
Name                           Value
----                           -----
dll                            False
xt                             cmake
d                              C:\testingGame\HelloCpp
p                              winuwp
xc                             {}
cc
winsdk
prefix                         C:\axmol\tools\external
setupOnly                      False
xb                             {--target, HelloCpp, --config, Debug}
a                              x64



build1k: proj_dir=C:\testingGame\HelloCpp, prefix=C:\axmol\tools\external
build1k: Found suitable installed glslcc: C:\axmol\tools\external\glslcc\glslcc.exe, version: 1.7.6
build1k: Found suitable installed cmake: C:\Program Files\CMake\bin\cmake.exe, version: 3.27.0
build1k: Building target winuwp on windows with toolchain msvc ...
build1k: CONFIG_ALL_OPTIONS=-DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0 -A x64, Count=4
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.
CMake Error at CMakeLists.txt:30 (project):
  Failed to run MSBuild command:

    C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe

  to get the value of VCTargetsPath:

    MSBuild version 17.5.0+6f08c67f3 for .NET Framework
    Build started 7/9/2023 9:13:15 PM.
    Included response file: C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\MSBuild.rsp

    Project "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" on node 1 (default targets).
    C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(832,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj]
    Done Building Project "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" (default targets) -- FAILED.

    Build FAILED.

    "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" (default target) (1) ->
    (_CheckForInvalidConfigurationAndPlatform target) ->
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(832,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj]

        0 Warning(s)
        1 Error(s)

    Time Elapsed 00:00:00.07


  Exit code: 1



-- Configuring incomplete, errors occurred!
build1k: BUILD_ALL_OPTIONS=--target HelloCpp --config Debug --parallel, Count=5
MSBuild version 17.5.0+6f08c67f3 for .NET Framework
MSBUILD : error MSB1009: Project file does not exist.
Switch: HelloCpp.vcxproj
PS C:\testingGame\HelloCpp>```

**Reply by aismann** [reply]
> build1k: PowerShell 5.1.19041.3031

I got also probs with  PowerShell 5.1.19041.3031

Please use the this one:
![image](https://github.com/axmolengine/axmol/assets/8652787/2c427eb1-e25f-4bdb-b971-37f7254712c3)

**Reply by rudiHammad** [reply]
Ok, I installed powershell-7 and still get the same error

```PS C:\testingGame\HelloCpp> .\build.ps1 -p winuwp
build1k: PowerShell 7.3.5
build1k:
Name                           Value
----                           -----
winsdk
setupOnly                      False
xb                             {--target, HelloCpp, --config, Debug}
p                              winuwp
prefix                         C:\axmol\tools\external
cc
a                              x64
dll                            False
d                              C:\testingGame\HelloCpp
xt                             cmake
xc                             {}


build1k: proj_dir=C:\testingGame\HelloCpp, prefix=C:\axmol\tools\external
build1k: Found suitable installed glslcc: C:\axmol\tools\external\glslcc\glslcc.exe, version: 1.7.6
build1k: Found suitable installed cmake: C:\Program Files\CMake\bin\cmake.exe, version: 3.27.0
build1k: Building target winuwp on windows with toolchain msvc ...
build1k: CONFIG_ALL_OPTIONS=-DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0 -A x64, Count=4
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.
CMake Error at CMakeLists.txt:30 (project):
  Failed to run MSBuild command:

    C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe

  to get the value of VCTargetsPath:

    MSBuild version 17.5.0+6f08c67f3 for .NET Framework
    Build started 7/9/2023 9:32:56 PM.
    Included response file: C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\MSBuild.rsp

    Project "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" on node 1 (default targets).
    C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(832,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj]
    Done Building Project "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" (default targets) -- FAILED.

    Build FAILED.

    "C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj" (default target) (1) ->
    (_CheckForInvalidConfigurationAndPlatform target) ->
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(832,5): error : The BaseOutputPath/OutputPath property is not set for project 'VCTargetsPath.vcxproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='x64'.  You may be seeing this message because you are trying to build a project without a solution file, and have specified a non-default Configuration or Platform that doesn't exist for this project. [C:\testingGame\HelloCpp\build_x64\CMakeFiles\3.27.0-rc4\VCTargetsPath.vcxproj]

        0 Warning(s)
        1 Error(s)

    Time Elapsed 00:00:00.08


  Exit code: 1



-- Configuring incomplete, errors occurred!
build1k: BUILD_ALL_OPTIONS=--target HelloCpp --config Debug --parallel, Count=5
MSBuild version 17.5.0+6f08c67f3 for .NET Framework
MSBUILD : error MSB1009: Project file does not exist.
Switch: HelloCpp.vcxproj
PS C:\testingGame\HelloCpp>```

**Comment by aismann** [comment]
Have you run `setup.ps1` before?
Please have a look on "axmol\docs\DevSetup.md" too

**Comment by aismann** [comment]
@halx99 
Small whish:
`pwsh .\build.ps1` should be create a `build_x64` folder
`pwsh .\build.ps1 -p winuwp` should be create a `build_uwp` folder
Both should be still there after creating.

**Reply by halx99** [reply]
yes, by default, should create a different build folder since we can build multi target platforms on windows: win32, winuwp, android

otherwise, if you build them one by one without folder deleting, will cause cmake generate error.

**Reply by halx99** [reply]
The latest commit: https://github.com/axmolengine/axmol/commit/09a3a950ae77d658758bc678c923a96929fe1b57 improve build folder create, the default build folder name rules:

- host target: build_${buildArch}
- cross target: build_${buildTarget}_{buildArch}

**Comment by rudiHammad** [comment]
yes I have been looking at "axmol\docs\DevSetup.md" to follow the guide. I did go into c:\axmol and run setup.ps1.
Android build is fine. For this one I don't know what I am missing, I am sure I did everything is in the guide.
The error says:
`Failed to run MSBuild command:` rings any bells?

**Reply by aismann** [reply]
I start alwyas on a developer command shell:
![image](https://github.com/axmolengine/axmol/assets/8652787/d1651a02-c614-453f-863e-21bd2fa069ed)

**Reply by rudiHammad** [reply]
I dont think that make any difference. Just to be clear once again in case I am not explainin my self properly.
This works:
`pwsh build.ps1 -a x64`
`pwsh build.ps1 -a x86`

This does not work with the errors mentioned before.
`pwsh .\build.ps1 -p winuwp`
`pwsh .\build.ps1 -p win32`

I am not sure what else to do to troubleshoot this.

**Comment by halx99** [comment]
`pwsh .\build.ps1 -p win32` same with `pwsh build.ps1 -a x64` on windows host

**Comment by aismann** [comment]
I tried to start cpp_tests	
![image](https://github.com/axmolengine/axmol/assets/8652787/edba353e-11cc-48ad-939c-904380d69e97)

And got this errror :
```
Creating a new clean layout...
Copying files: Total 95 mb to layout...
Checking whether required frameworks are installed...
Registering the application to run from layout...
DEP0700: Registration of the app failed. [0x80070005] 
```

`Error		DEP0700: Registration of the app failed. [0x80070005]	cpp_tests		`

**Reply by halx99** [reply]
Try 
![image](https://github.com/axmolengine/axmol/assets/6977319/b21b815a-dca2-440c-98b6-95e68df755d2)

**Comment by rudiHammad** [comment]
hey guys,
so I was able to send the .exe file demo to some friends to test. All I did is copy the release folder that created visual studio created, and send it as it is.
![Capture](https://github.com/axmolengine/axmol/assets/25636976/240eb000-f1f0-4ad4-a1f6-7421f559c50c)
And all they had to do is  execute the .exe and it worked.

I have never published any game, I am very new to this, so can you explain how do you deploy a game in itch or steam.
I am not talking about having to register, download steam sdk for achievements and stuff... I am not there yet.
I am asking about the very basic. Would it be as easy as uploading the files from the folder in the image above to the distributing platform?

**Comment by halx99** [comment]
What about nsis for windows, engine itself not provide package to one image support, and vcredist required on machine which visual studio(version should match with your build machine) not installed.

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by rudiHammad** [reply]
Interesting. I'll check them out, thanks.
Going a bit off topic here, but has anyone integrated axmol with steam sdk?
https://partner.steamgames.com/doc/sdk

**Comment by rudiHammad** [comment]
Finally! it is all working! Those where my mistakes.
1. my path had spaces 
2. I created the visual studio project with `cmake -S . -B build -G "Visual Studio 17 2022" -A x64`
 doing that was making the commands ` pwsh .\build.ps1 -p winuwp` and ` pwsh .\build.ps1 -p win32` error.
But if I build the visual studio project with `pwsh build.ps1 -p win32` then those commands worked perfectly.

One thing that I don't understand though is that if I run this: `pwsh build.ps1 -p win32 --config 'Release'`
it still builds the files in the debug folder. Insn't it supposes to build it in a relase folder?

**Comment by halx99** [comment]
> 1. my path had spaces

1. it's seems cmake issue, I just create project name contains spaces, run `cmake -B build`, the cmake generate will failed with:
```txt
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
CMake Error: Could not find cmake module file: CMakeDetermineGameCompiler.cmake
CMake Error: Could not find cmake module file: D:/dev/projects/Example Game/build/CMakeFiles/3.27.0-rc3/CMakeGameCompiler.cmake
CMake Error at CMakeLists.txt:30 (project):
  No CMAKE_Game_COMPILER could be found.



CMake Error: Could not find cmake module file: CMakeGameInformation.cmake
-- Configuring incomplete, errors occurred!
```

**Reply by rudiHammad** [reply]
yup.

**Comment by rudiHammad** [comment]
Okey, i need some help understanding how to send the build on windows for people to tested. This is what I have done:
1. built the default axmol test using `pwsh build.ps1 -p win32`, no errors, all good this time.
2. copied the content in into a folder and compressed it as .zip. This is the content
![res](https://github.com/axmolengine/axmol/assets/25636976/e9cf4424-bb7b-410c-85f5-c67c0af8d376)

3. used NSIS to create an installer
4. send the installer to people.

The result is that for half of the people I've sent it, it worked with no problems.
And The other half reported the errors bellow when they tried to run GameTest.exe
![1](https://github.com/axmolengine/axmol/assets/25636976/1530051d-2c57-4b47-be78-a8ef26409ab4)
![4](https://github.com/axmolengine/axmol/assets/25636976/560827a1-5dad-486f-afe9-7ed58663ca55)
![2](https://github.com/axmolengine/axmol/assets/25636976/0c6b6180-f3c8-495c-8175-5886c51ebd48)
![3](https://github.com/axmolengine/axmol/assets/25636976/c41c7039-3d34-4e11-868a-9e12b7eadabe)

Can someone guide me through how you do send your builds on windows to people? Please assume I have never done that so don't assume that I am aware the basics. Even something as stupid as : step 1. turn on you computer. step 2, step 3...
I am currently on the edge on switching to unity, but I don't want to because I love C++ and Axmol. I just need to be sure that I am not gonna spend a year making a game that I can deploy.

At this point this thread as been a great help. Thank you for that. I just want to push through these last issues.
I understand the errors  are because those .dll are not present in the system, but I can't ask to players to download and install them.
It must be ready to go, and I don't know if I am doing something wrong.

**Reply by halx99** [reply]
1. You should execute vcredist install in your nsis install setup script, download from: https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170, just match with your build arch.

**Reply by halx99** [reply]
Anothor choice what I'm using in `x-studio ide` is `msvcr14x`: https://github.com/sonyps5201314/msvcr14x

showcase: https://github.com/simdsoft/x-studio/releases/tag/v2022.1.7

**Reply by halx99** [reply]
by the way, you should build release binaries, not debug

**Reply by rudiHammad** [reply]
yup I know. I did post above that I tried to build in Release using `pwsh build.ps1 -p win32 --config 'Release'`
but it still went into debug.
I'll try to do the vcredist install first.

**Reply by halx99** [reply]
the commandline incorret, try open powershell console window, then run:

`.\buile.ps1 -xb '--config','Release'`

**Reply by rudiHammad** [reply]
thanks. My bad.

**Comment by rudiHammad** [comment]
> You should execute vcredist install in your nsis install setup script, download from: https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170, just match with your build arch.

Okey, I did some google reasearch and ask chat gpt on how to do that. What I got is that after download VC_redist.x64.exe, which is my arch, I have to create the following script, which I named vcredistInstaller.nsi and it goes like this

```# Set the name of your installer
Outfile "MyInstaller.exe"

# Set the default installation directory
InstallDir $PROGRAMFILES\MyApp

# Default section
Section

    # Set output path to the installation directory
    SetOutPath $INSTDIR

    # Copy the files
    File "HelloCpp.exe"
    File "libcrypto-3-x64.dll"
    File "HelloCpp.pdb"
    File "libcurl.dll"
    File "libssl-3-x64.dll"
    File "OpenAL32.dll"
    File "zlib1.dll"

    # Copy the contents of the "Content" folder to the installation directory
    File /r "Content\*"

    # Execute the Visual C++ Redistributable installation
    ExecWait '"vcredist.exe" /install'

    # Launch HelloCpp.exe after the Visual C++ Redistributable installation is complete
    Exec '"$INSTDIR\HelloCpp.exe"'

SectionEnd```

the line ` ExecWait '"vcredist.exe" /install'` is i guess what you meant, but of course you have to do everything else too right?.. Anyway, I did that. Opened NSI, Drag and drop the script into Compile NSI scripts. It worked with no errors, which created  MyInstaller.exe. Executed that, all good. But again when I send it to others it didn't work. Same errors .dll errors.
In theory running that script should have included those missing dlls right?

I'd rather work in visual studio. I don't think changing to another idle we'll fix the issue. 

I guess I'll keep trying...

**Comment by halx99** [comment]
and vs console have command `dumpbin /dependents xxx.exe` to help check exe or dll dependencies

**Comment by rudiHammad** [comment]
Sorry man, I give up. I am just getting drops of information with every reply and I have to spend hours googling what it means. First it was learning what nsis was, then learning what  "vcredist install in your nsis install setup script" means, which is learning what a .nsi file, how to create one and running from the nsis. Now I have to figure how what "vs console have command dumpbin /dependents xxx.exe to help check exe or dll dependencies" means. 

I am having a lot of trouble following up with just drops of information. It is probably me. Thanks for all you help anyways.

R

**Comment by halx99** [comment]
it's ok, sorry the win32 package not part of axmol, axmol still a framework, not with full rich game develop tool. but I can confirm axmol should more easy to deploy win32 target than original cocos2dx, brcause all prebuilts dll is matched vcredist version aka msvc14.36.
and require developer have rich windows dev experience. I perfer use msvcr14x what I using in x-studio ide(also use axmol), it's publish microsoft app store, and other PC platforms.

**Reply by rudiHammad** [reply]
did you publish on steam too?

**Comment by halx99** [comment]
A easy way is copy all vcredist dlls to target bin directory, this can be done powershell build script in the future.

获取Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by DelinWorks** [comment]
I got the same problem as you, you see there are C++ runtime redistributebles that have to be installed on the system https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170, the way you are publishing your game is indeed correct, but unfortunately these programs need to be installed on the user's machine so you need to provide vcredist, there is a game called geometry dash that suffers from this problem and also unreal engine games too, you just need to google vcredist c++ x64 and download all execs for all dates (2008, 2013, 2017, and so on) that way the missing dlls are present in the system, you could add a folder to your game called, `vcredist` and people should install whatevers in that folder, also there is another method of publishing which is packing your game files into an installer, that installer will install your game on a specific path, create a shortcut on desktop for the user and run the important vcredist setup files, most games do it this way, also if you published your game through steam, steam will scan your game and see what libraries and runtimes it requires and automatically installs them for that game before launch, I hope that helps!

**Comment by rudiHammad** [comment]
ahhh, thanks a lot. I am gonna try that. I really don't what to go to Unity XD. I am sure it is cool, but I prefer to stay in C++ and axmol.
I did downlaod and installed VC_redist.x64.exe. But the moment you ask the user to do  the install of something else than the game you lose him, so the easier the better.
Would please explain exactly how to pack you game into the installer with all the vcredist setup files. I tried that using nsis through the script option which in theory added the vcredist, as halx99 suggested.

```Outfile "MyInstaller.exe"

# Set the default installation directory
InstallDir $PROGRAMFILES\MyApp

# Default section
Section

    # Set output path to the installation directory
    SetOutPath $INSTDIR

    # Copy the files
    File "HelloCpp.exe"
    File "libcrypto-3-x64.dll"
    File "HelloCpp.pdb"
    File "libcurl.dll"
    File "libssl-3-x64.dll"
    File "OpenAL32.dll"
    File "zlib1.dll"

    # Copy the contents of the "Content" folder to the installation directory
    File /r "Content\*"

    # Execute the Visual C++ Redistributable installation
    ExecWait '"vcredist.exe" /install'

    # Launch HelloCpp.exe after the Visual C++ Redistributable installation is complete
    Exec '"$INSTDIR\HelloCpp.exe"'

SectionEnd
```

I think that a guide for dummies (like my self apparently) would bring more people.
Just if someone can create the default Helloworld project, and go through the process of creating the installer and the vcredist dependencies, would help a lot. With images with big huge red arrows guiding step by step XD.
I am not asking anyone to do the work for me, and proof of it is that I have been trying for days now. But I don't know where I am going wrong anymore. If I succed I will be happy to contribute making that guide for axmol github my self.
I think that many of us are interested on Pc mainly (basically steam), so the easier we make it, more people will come to axmol.

**Reply by DelinWorks** [reply]
could you move this discussion to the discord unofficial axmol server, help is way easier there

**Reply by rudiHammad** [reply]
Sure

**Comment by rudiHammad** [comment]
ok, done. Here is the discord link
https://discord.com/channels/1099599084895088670/1128369237623312525
We can close it discussion here and continue there. It has gone off topic anyways.
Thanks for the replies.
